### PR TITLE
Augmented Neural DE

### LINKED ISSUE #288: Augmented neural ODEs
https://arxiv.org/abs/1904.01681

### PR DESCRIPTION
**Pull request by avik-pal**
Fixes #288.

TODOs before it is ready for review:
- [x] Minimal Example of Augmented ODE
- [x] Test working with the other DE layers
- [x] Documentation
- [x] Unit Tests
- [x] More details in the example

### CODE REVIEW

File: docs/src/examples/Augmented-NN-ODE.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+# Expected Output
+
+```julia
+┌ Info: Generating Dataset
+└ @ Main In[11]:60
+┌ Info: Training Neural ODE
+└ @ Main In[11]:75
+┌ Info: Epoch 1
+└ @ Main /home/avik-pal/.julia/packages/Flux/Fj3bt/src/optimise/train.jl:121
```
Comment (ChrisRackauckas) [inline docs/src/examples/Augmented-NN-ODE.md:110]: maybe we don't want to do it as `@info` to get rid of the line number?

**Comment by avik-pal** [comment]
I added some tests, but not sure why the CI is not triggered.

File: test/runtests.jl [hunk 2]
```diff
@@ -24,6 +24,7 @@ end
 
 if GROUP == "All" || GROUP == "DiffEqFlux" || GROUP == "BasicNeuralDE"
     @safetestset "Neural DE Tests" begin include("neural_de.jl") end
+    @safetestset "Augmented Neural DE Tests" begin include("fast_neural_ode.jl") end
```
Comment (Vaibhavdixit02) [inline test/runtests.jl:27]: Looks like a wrong file is used for the test here?

File: src/neural_de.jl [hunk 3]
```diff
@@ -470,3 +470,37 @@ function (n::NeuralODEMM{M})(x,p=n.p) where {M<:FastChain}
...
+    nde::DE
+    adim::Int
+end
+
+Flux.@functor AugmentedNDELayer (nde,)
+
+(ande::AugmentedNDELayer)(x, args...) = ande.nde(augment(x, ande.adim), args...)
+
+augment(x::AbstractVector{S}, augment_dim::Int) where S =
+    cat(x, zeros(S, (augment_dim,)), dims = 1)
```
Comment (ChrisRackauckas) [inline src/neural_de.jl:500]: I see, yes because the augmented dimensions have null dynamics this works.

**Comment by avik-pal** [comment]
The test failures seem to be from something unrelated to the Augmentation Layer

**Pull request merged**